src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -2964,9 +2963,21 @@ private RefKind GetEffectiveParameterRefKind(ParameterSymbol parameter, RefKind
                 }
                 else
                 {
+                    RefKind argumentRefKind = arguments.RefKind(argumentPosition);
                     RefKind parameterRefKind = parameters.ParameterRefKinds.IsDefault ? RefKind.None : parameters.ParameterRefKinds[argumentPosition];
                     bool forExtensionMethodThisArg = arguments.IsExtensionMethodThisArgument(argumentPosition);
 
+                    if (forExtensionMethodThisArg)
+                    {
+                        Debug.Assert(argumentRefKind == RefKind.None);
+                        if (parameterRefKind == RefKind.Ref)
```

Comment:
Will add a comment explaining why we don't want to do that for other types.

---
In reply to: [122018204](https://github.com/dotnet/roslyn/pull/19703#discussion_r122018204) [](ancestors = 122018204,122015601)